web3/providers/async_ipc.py

```diff
@@ -0,0 +1,166 @@
+    RPCResponse,
+)
+
+from .async_base import (
+    AsyncJSONBaseProvider,
+)
+from .ipc import (
+    get_default_ipc_path,
+    has_valid_json_rpc_ending,
+)
+
+
+async def async_get_ipc_socket(
+    ipc_path: str, timeout: float = 2.0
+) -> Tuple[asyncio.StreamReader, asyncio.StreamWriter]:
+    if sys.platform == "win32":
+        # On Windows named pipe is used. Simulate socket with it.
+        from web3._utils.windows import (
+            NamedPipe,
+        )
+
+        return NamedPipe(ipc_path)
+    else:
+        return await asyncio.open_unix_connection(ipc_path)
+
+
+class PersistantSocket:
+    sock: Optional[Tuple[asyncio.StreamReader, asyncio.StreamWriter]] = None
+
+    def __init__(self, ipc_path: str) -> None:
+        self.ipc_path = ipc_path
+
+    async def __aenter__(self) -> Tuple[asyncio.StreamReader, asyncio.StreamWriter]:
+        if not self.ipc_path:
+            raise FileNotFoundError(
+                f"cannot connect to IPC socket at path: {self.ipc_path!r}"
+            )
+
+        if not self.sock:
+            self.sock = await self._open()
+        return self.sock
+
+    async def __aexit__(
+        self,
+        exc_type: Type[BaseException],
+        exc_value: BaseException,
+        traceback: TracebackType,
+    ) -> None:
+        # only close the socket if there was an error
+        if exc_value is not None:
+            try:
+                reader, writer = self.sock
+                writer.close()
+                await writer.wait_closed()
+            except Exception:
+                pass
+            self.sock = None
+
+    async def _open(self) -> Tuple[asyncio.StreamReader, asyncio.StreamWriter]:
+        return await async_get_ipc_socket(self.ipc_path)
+
+    async def reset(self) -> Tuple[asyncio.StreamReader, asyncio.StreamWriter]:
+        reader, writer = self.sock
+        writer.close()
+        await writer.wait_closed()
+        self.sock = await self._open()
+        return self.sock
+
+
+_async_session_cache_lock = threading.Lock()
+_async_session_pool = ThreadPoolExecutor(max_workers=1)
+
+
+class AsyncIPCProvider(AsyncJSONBaseProvider):
+    logger = logging.getLogger("web3.providers.AsyncIPCProvider")
+    _socket = None
+
+    def __init__(
+        self,
+        ipc_path: Optional[Union[str, Path]] = None,
+        timeout: int = 10,
+    ) -> None:
+        if ipc_path is None:
+            self.ipc_path = get_default_ipc_path()
+        elif isinstance(ipc_path, str) or isinstance(ipc_path, Path):
+            self.ipc_path = str(Path(ipc_path).expanduser().resolve())
+        else:
+            raise TypeError("ipc_path must be of type string or pathlib.Path")
+
+        self.timeout = timeout
+        self._socket = PersistantSocket(self.ipc_path)
+        super().__init__()
+
+    def __str__(self) -> str:
+        return f"<{self.__class__.__name__} {self.ipc_path}>"
+
+    async def make_request(self, method: RPCEndpoint, params: Any) -> RPCResponse:
```

Comment:
This question builds a bit on the comment from @SheldonHolmgren. If we create a reader and writer here each time we make a request, we aren't exactly utilizing the "persistent open socket" paradigm in a way that could be advantageous to state preservation requests like subscriptions.

I'm curious how much work it may take to make this a `PersistentConnectionProvider` or something similar to where we can keep an open connection throughout the life of the provider rather than opening / closing each time. It may take more work and breaking changes to try to come back later and put that work in.

Not necessarily trying to head in that direction as that may take quite a bit more work, but curious on everyone's thoughts on that.

Reply:
Yeah, I was thinking about that too. IMHO it makes sense to allow for `eth_subscribe` support here, but not sure how much work it'll take. I'll timebox it and if I can get something working, I'll add in a new commit for reviewing.

Reply:
Sounds good. In a best case scenario, assuming we go with the message listener task that's currently in PR for the `WebsocketProviderV2`, I think we can probably hijack all of the `RequestProcessor` logic, switch the naming within the more broad `PersistentConnectionProvider` class and the `RequestManager` (from `ws` or `websocket` to `socket` maybe?) while using the appropriate logic for the specific sockets in the implementation classes. We can likely route the `socket.reader` into a message listener background task in a similar way as iterating over new messages on the websocket is done in wsv2. Curious to see what you find out as you timebox it though 👀